### PR TITLE
Implement `GetScExecutionEvents` in gRPC API

### LINKED ISSUE #3867: Implement `GetScExecutionEvents` in gRPC API
This method will be a part of the new gRPC API as it's needed to be used primarily in newer Massa architecture.

We should improve the input/output of this function compared to the existing one and make It straightforward. 

Related to https://github.com/massalabs/massa/discussions/3406

### PR DESCRIPTION
**Pull request by aoudiamoncef**
* [x] document all added functions
* [ ] try in sandbox /simulation/labnet
* [x] unit tests on the added/changed features
  * [x] make tests compile
  * [x] make tests pass 
* [x] add logs allowing easy debugging in case the changes caused problems
* [x] if the API has changed, update the API specification

closes #3867

### Summary
🚀🛠️📝

This pull request adds a new gRPC method `GetScExecutionEvents` to query smart contract execution events from the node. It also introduces a new configuration parameter `max_operation_ids_per_request` to limit the size of gRPC requests. The changes involve updating the `massa-grpc`, `massa-models`, `massa-proto`, and `massa-node` crates, as well as the documentation files.

> _`GetScExecutionEvents`_
> _gRPC method for smart contracts_
> _autumn leaves fall fast_

### Walkthrough
*  Add a new gRPC method `GetScExecutionEvents` to query smart contract execution events from the execution layer ([link](https://github.com/massalabs/massa/pull/3868/files?diff=unified&w=0#diff-966d81c5c17e23841cc02b5985aec0006e8cdf46d3da3e9af7d1037a204a3868R72-R79), [link](https://github.com/massalabs/massa/pull/3868/files?diff=unified&w=0#diff-798542cd089e9e67e2cab48d3814caacfc251213c60a050e968c493ea39a5062R1480-R1507), [link](https://github.com/massalabs/massa/pull/3868/files?diff=unified&w=0#diff-798542cd089e9e67e2cab48d3814caacfc251213c60a050e968c493ea39a5062R1909-R1916), [link](https://github.com/massalabs/massa/pull/3868/files?diff=unified&w=0#diff-798542cd089e9e67e2cab48d3814caacfc251213c60a050e968c493ea39a5062R2429-R2474), [link](https://github.com/massalabs/massa/pull/3868/files?diff=unified&w=0#diff-a823bebc760b940b7f058efcd7874fa5a39e55f206e754197c87d84f92aea155R1207), [link](https://github.com/massalabs/massa/pull/3868/files?diff=unified&w=0#diff-87c1cb3b00b9f46af18c6a640940122841604d9dade89ec535a345b106bae52aR2727-R2733), [link](https://github.com/massalabs/massa/pull/3868/files?diff=unified&w=0#diff-87c1cb3b00b9f46af18c6a640940122841604d9dade89ec535a345b106bae52aR2898-R2907))
*  Define new messages and enum for the request and response types and the event status of the `GetScExecutionEvents` method in the `api.proto` file ([link](https://github.com/massalabs/massa/pull/3868/files?diff=unified&w=0#diff-966d81c5c17e23841cc02b5985aec0006e8cdf46d3da3e9af7d1037a204a3868R329-R406))
*  Generate new structs and enum for the request and response types and the event status of the `GetScExecutionEvents` method in the `massa.api.v1.rs` file using the `prost` and `tonic` crates ([link](https://github.com/massalabs/massa/pull/3868/files?diff=unified&w=0#diff-798542cd089e9e67e2cab48d3814caacfc251213c60a050e968c493ea39a5062R687-R784), [link](https://github.com/massalabs/massa/pull/3868/files?diff=unified&w=0#diff-798542cd089e9e67e2cab48d3814caacfc251213c60a050e968c493ea39a5062R1143-R1181))
*  Implement conversions between the gRPC and the internal types for the request and response types and the event status of the `GetScExecutionEvents` method in the `mapping_grpc.rs` file using the `From` and `TryFrom` traits ([link](https://github.com/massalabs/massa/pull/3868/files?diff=unified&w=0#diff-8d388876bf9b95596cca493e9b26d724994f068637fdf817b4de58494791e6aaR219-R283))
*  Add a new function `get_sc_execution_events` to the `api.rs` file that takes a gRPC request and returns a gRPC response with the smart contract execution events that match the given filter ([link](https://github.com/massalabs/massa/pull/3868/files?diff=unified&w=0#diff-7e3c322726af7157e1a41afc4163e23255520ca33abc70209f9ff4a895b3cbcbR495-R542))
*  Add a new handler function `get_sc_execution_events` to the `handler.rs` file that calls the `get_sc_execution_events` function from the `api.rs` file and wraps the result in a tonic response ([link](https://github.com/massalabs/massa/pull/3868/files?diff=unified&w=0#diff-8d24702ba466d12fc108ff65a78c43efedcbd949f7c1427a16b176861239dcd6R75-R84))
*  Add a new configuration parameter `max_operation_ids_per_request` to the `config.toml` file and the `GrpcConfig` struct that sets the maximum number of operation ids that can be included in a single gRPC request ([link](https://github.com/massalabs/massa/pull/3868/files?diff=unified&w=0#diff-5275c26ef00940f5cb65a70bbaac1ffe9e40f1fd639b1bec4e7ae9f131972184R76-R77), [link](https://github.com/massalabs/massa/pull/3868/files?diff=unified&w=0#diff-1f883ba25a829313ad70daa6c31595706cf842a63e6e7bc0c342d89081084226R288-R289))
*  Add the `max_operation_ids_per_request` field to the `MassaGrpc` struct initialization in the `main.rs` file and use it in the error message of the `get_sc_execution_events` function in the `api.rs` file ([link](https://github.com/massalabs/massa/pull/3868/files?diff=unified&w=0#diff-535b75052259fc1e2f15ad89ecd55326f762d8a118074d573377d09116533761R767), [link](https://github.com/massalabs/massa/pull/3868/files?diff=unified&w=0#diff-7e3c322726af7157e1a41afc4163e23255520ca33abc70209f9ff4a895b3cbcbL374-R375), [link](https://github.com/massalabs/massa/pull/3868/files?diff=unified&w=0#diff-7e3c322726af7157e1a41afc4163e23255520ca33abc70209f9ff4a895b3cbcbL387-R389))
*  Add new import statements to use the new types and functions related to the `GetScExecutionEvents` method in the `api.rs`, `handler.rs`, and `mapping_grpc.rs` files ([link](https://github.com/massalabs/massa/pull/3868/files?diff=unified&w=0#diff-7e3c322726af7157e1a41afc4163e23255520ca33abc70209f9ff4a895b3cbcbR9), [link](https://github.com/massalabs/massa/pull/3868/files?diff=unified&w=0#diff-8d24702ba466d12fc108ff65a78c43efedcbd949f7c1427a16b176861239dcd6L7-R8), [link](https://github.com/massalabs/massa/pull/3868/files?diff=unified&w=0#diff-8d388876bf9b95596cca493e9b26d724994f068637fdf817b4de58494791e6aaL3-R11))
*  Add new sections and entries to the `api.md` and `index.html` files to document the new types and functions related to the `GetScExecutionEvents` method ([link](https://github.com/massalabs/massa/pull/3868/files?diff=unified&w=0#diff-a823bebc760b940b7f058efcd7874fa5a39e55f206e754197c87d84f92aea155R30-R34), [link](https://github.com/massalabs/massa/pull/3868/files?diff=unified&w=0#diff-a823bebc760b940b7f058efcd7874fa5a39e55f206e754197c87d84f92aea155R59-R60), [link](https://github.com/massalabs/massa/pull/3868/files?diff=unified&w=0#diff-a823bebc760b940b7f058efcd7874fa5a39e55f206e754197c87d84f92aea155R74), [link](https://github.com/massalabs/massa/pull/3868/files?diff=unified&w=0#diff-a823bebc760b940b7f058efcd7874fa5a39e55f206e754197c87d84f92aea155R487-R569), [link](https://github.com/massalabs/massa/pull/3868/files?diff=unified&w=0#diff-a823bebc760b940b7f058efcd7874fa5a39e55f206e754197c87d84f92aea155R944-R979), [link](https://github.com/massalabs/massa/pull/3868/files?diff=unified&w=0#diff-a823bebc760b940b7f058efcd7874fa5a39e55f206e754197c87d84f92aea155R1175-R1188), [link](https://github.com/massalabs/massa/pull/3868/files?diff=unified&w=0#diff-87c1cb3b00b9f46af18c6a640940122841604d9dade89ec535a345b106bae52aR274-R293), [link](https://github.com/massalabs/massa/pull/3868/files?diff=unified&w=0#diff-87c1cb3b00b9f46af18c6a640940122841604d9dade89ec535a345b106bae52aR390-R397), [link](https://github.com/massalabs/massa/pull/3868/files?diff=unified&w=0#diff-87c1cb3b00b9f46af18c6a640940122841604d9dade89ec535a345b106bae52aL418-R451), [link](https://github.com/massalabs/massa/pull/3868/files?diff=unified&w=0#diff-87c1cb3b00b9f46af18c6a640940122841604d9dade89ec535a345b106bae52aR1293-R1468), [link](https://github.com/massalabs/massa/pull/3868/files?diff=unified&w=0#diff-87c1cb3b00b9f46af18c6a640940122841604d9dade89ec535a345b106bae52aR2143-R2232), [link](https://github.com/massalabs/massa/pull/3868/files?diff=unified&w=0#diff-87c1cb3b00b9f46af18c6a640940122841604d9dade89ec535a345b106bae52aR2637-R2671))

### CODE REVIEW

File: massa-proto/proto/massa/api/v1/api.proto [hunk 1]
```diff
@@ -318,6 +326,84 @@ message OperationsContext {
...
+  // Sc execution context
+  ScExecutionEventContext context = 1;
+  // json data string
+  string data = 2;
+}
+
+// ScExecutionEvent context
+message ScExecutionEventContext {
+  // When was it generated
+  Slot origin_slot = 2;
```
Comment (modship) [inline massa-proto/proto/massa/api/v1/api.proto:386]: why you doesn"t use context ?

File: massa-grpc/src/config.rs [hunk 2]
```diff
@@ -86,4 +86,6 @@ pub struct GrpcConfig {
     pub max_denunciations_per_block_header: u32,
     /// max number of block ids that can be included in a single request
     pub max_block_ids_per_request: u32,
+    /// max number of operation ids that can be included in a single request
+    pub max_operation_ids_per_request: u32,
```
Comment (modship) [inline massa-grpc/src/config.rs:90]: that's big, no ?

**Review by modship — APPROVED** [review]

File: massa-proto/proto/massa/api/v1/api.proto (hunk 1 above)
Comment (aoudiamoncef) [reply massa-proto/proto/massa/api/v1/api.proto:386]: The context in the top is the query execution context

This context is the SC call execution context

File: massa-grpc/src/config.rs (hunk 2 above)
Comment (aoudiamoncef) [reply massa-grpc/src/config.rs:90]: We have a lot of similare settings which uses u32, so

File: massa-proto/proto/massa/api/v1/api.proto (hunk 1 above)
Comment (modship) [reply massa-proto/proto/massa/api/v1/api.proto:386]: Oh yes ok, and this slot is the slot when the sc was generated ?  You can"t use context for that ?

Comment (modship) [reply massa-proto/proto/massa/api/v1/api.proto:386]: ok :)

**Pull request merged**